rfcs/0051-hard-fork-package-generation.md

```diff
@@ -0,0 +1,148 @@
+
+* Hard fork initialization
+  * When a daemon is configured with a genesis timestamp after the current time, it sleeps until genesis.
+  * When a daemon is configured with genesis state ledger hashes only (no accounts in config), it can successfully load those ledger from disk, assuming they have been packaged with the software.
+* Planned hard fork
+  * Daemons are able to dump ledger data without long async cycles or increased rates of validation timeout errors.
+  * Daemons are able to dump ledgers containing up to 1 million accounts (there is currently a bug with this due to RPC size limits)
+  * Infrastructure for deploying nodes is guaranteed to dump the required genesis ledger states.
+    * Such a test should run a few nodes, and occasionally kill various nodes. We want to ensure this system has high reliability.
+  * Buildkite pipeline is able to produce hard fork packages within the allowed time window (6 hours) even for large ledgers.
+    * We will use a multiple of the current mainnet ledger size when performing this test to ensure that the system meets requirements for future ledger sizes that we may have to perform unplanned hard forks from.
+* Unplanned hard fork
+  * The replayer tool is capable of materializing staged ledgers, snarked ledgers, and epoch ledgers from arbitrary blocks.
+  * The replayer tool is able to materialize any ledger at an arbitrary block within 4 hours.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+This approach does not currently include a plan for migrating the scan state data. This means that there will be some transaction history missing from the final blockchain proof of the prior network. It would be an additional and non-trivial engineering project to actually extend this proof retroactively to cover the missing transactions that we are not migrating into the new proof system. This leads to a weakening of Mina's trustlessness, and adds new requirements onto clients that want to truly interact with the Mina protocol's state in a trustless fashion.
+
+## Rationale and alternatives
+
+* ALTERNATIVE TO CLI-FLAG FOR DATA DUMPS: cron job that runs a script against graphql api to find candidates, then dumps via CLI
+* ALTERNATIVE TO PRE-DUMPING DATA: keep node alive after the network stop slot, and dump the data at that point
+    * this requires putting the node into a new state, and introduces new risk since the node could crash in that state and lose data
+
+## Unresolved questions
+
+* For emergency hard forks, do we need to consider the world in which we only take the snarked ledger an not the staged ledger?
+* Is JSON actually the best format for dumping, migrating, and loading ledgers? It's slow to serialize/deserialize, the human readability isn't important for this scope, and it has the potential to introduce data translation errors in ways that other serialization formats don't. Should we just use a `bin_prot` representation instead?
```

Comment:
That's a good question, which I also discuss a little in [PR #14213](https://github.com/MinaProtocol/mina/pull/14213). The advantage of this solution is its simplicity. It doesn't require any additional work. Is this also the case for `bin_prot` representation? Feel free to add relevant details in the "Rationale and alternatives" section in that RFC.

Also I disagree that human readability is unimportant. Because the community has to trust us on preparing the genesis ledger correctly, we should make it as easy for them as possible to check whether their account balances are right, ideally before the new network starts.

Reply:
Some way to visualize the format is important for humans, but the JSON format is rather wasteful in how it converts bignumbers into b64 encoded strings. It makes the size of the data very large and makes it take a while to process. But with the current ledger size, it's ok to keep it in JSON for now, I think. Using bin_prot would cut down the file size by about 1/4-1/5.